### PR TITLE
Refactor STM32F0xx dtsi

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR reorganizes the successive dtsi inclusion of the STM32F0 series to better reflect the subfamilies (STM32F0x0 on one side, STM32F0x1, x2 and x8 on the other). It allows to fix some errors of peripherals included in devices that do not have them, and to remove some duplication by having a more logical and clear inclusion tree.

Also add support for STM32F042 and Nucleo F042K6.

### CODE REVIEW

File: boards/arm/nucleo_f042k6/nucleo_f042k6.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
+ram: 6
+flash: 32
+supported:
+  - i2c
+  - spi
+  - gpio
```
Comment (erwango) [inline boards/arm/nucleo_f042k6/nucleo_f042k6.yaml:14]: adc

File: boards/arm/nucleo_f042k6/nucleo_f042k6.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+identifier: nucleo_f042k6
+name: ST Nucleo F042K6
+type: mcu
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
+ram: 6
```
Comment (erwango) [inline boards/arm/nucleo_f042k6/nucleo_f042k6.yaml:9]: On some other devices with limited available RAM, gpio d and f where disabled.
Is it safe to keep them?
Did you run compilation campaign on this board to detect test with potential RAM overflow ?

File: dts/arm/st/f0/stm32f091.dtsi [hunk 3]
```diff
@@ -4,73 +4,60 @@
...
-			interrupts = <15 0>;
-			interrupt-names = "global";
-			st,prescaler = <0>;
+		/*
+		 * USARTs 3-8 share the same IRQ on stm32f091xx devices. This
+		 * configuration is not currently supported, so at most one of
+		 * these may be enabled at a time. Enabling more than one will
+		 * result in a build failure.
+		 */
+		usart5: serial@40005000 {
```
Comment (erwango) [inline dts/arm/st/f0/stm32f091.dtsi:17]: This looks like an addition rather than a reorganization. A dedicated commit is required to ease the review.

**Review by erwango — CHANGES_REQUESTED** [review]
First commit would require to be split to ease review.
This is specially required if new nodes are added.

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**